File: Cores/ExtremumScalaCore/src/main/scala/OSOL/Extremum/Core/Scala/Vectors/IntervalVector.scala
package OSOL.Extremum.Core.Scala.Vectors

import OSOL.Extremum.Core.Scala.Arithmetics.Interval
import IntervalVector.Converters._
import OSOL.Extremum.Core.Scala.CodeFeatures.Pipe
import OSOL.Extremum.Core.Scala.Optimization.Optimizable
import OSOL.Extremum.Core.Scala.Vectors.Exceptions.DifferentKeysException

/** Interval valued vector
  *
  * @param values values which form VectorObject (key-value pairs)
  */
class IntervalVector private (override val values: Map[String, Interval])
  extends VectorObject[Interval](values) with Optimizable[IntervalVector, Interval] {

  final def equalsTo(that: VectorObject[Interval]): Boolean = {
    val keys_1 = this.keys
    val keys_2 = that.keys
    if (keys_1 != keys_2) throw new DifferentKeysException(keys_1, keys_2)
    else keys_1.forall(k => this(k) == that(k))
  }

  final override def add(that: VectorObject[Interval]): IntervalVector =
    this.elementWiseOp(that, (a, b) => a + b)

  final override def addImputeMissingKeys(that: VectorObject[Interval]): IntervalVector =
    this.elementWiseOpImputeMissingKeys(that, (a, b) => a + b, Interval(0.0))

  final override def multiply(that: VectorObject[Interval]): IntervalVector =
    this.elementWiseOp(that, (a, b) => a * b)

  final override def multiplyImputeMissingKeys(that: VectorObject[Interval]): IntervalVector =
    this.elementWiseOpImputeMissingKeys(that, (a, b) => a * b, Interval(1.0))

  final override def multiply(coefficient: Double): IntervalVector =
    this.keys.map(k => (k, Interval(coefficient, coefficient) * this (k))) |> IntervalVector.apply

  final override def moveBy(delta: (String, Double)*): IntervalVector = {
    val deltaWithImputedValues = (delta ++ this.keys.diff(delta.map(_._1).toSet).map(k => (k, 0.0)))
      .map { case (k, v) => (k, Interval(v))}
    this.add(deltaWithImputedValues |> IntervalVector.apply)
  }

  final override def constrain(area: (String, (Double, Double))*): IntervalVector = {
    def constrainValue(value: Double, restrictingArea: (Double, Double)): Double = {
      val (min, max) = restrictingArea
      if (value > max) max
      else {
        if (value < min) min
        else value
      }
    }
    val restrictingArea = area.toMap
    val constrainedVector = this.values.map { case (k, value) => (k,
      k match {
        case _ if restrictingArea.isDefinedAt(k) =>
          Interval(constrainValue(value.lowerBound, restrictingArea(k)), constrainValue(value.upperBound, restrictingArea(k)))
        case _ => value
      })
    }
    constrainedVector
  }

  final override def getPerformance(f: Map[String, Interval] => Interval): Double = f(this.values).lowerBound

  final override def toBasicForm(): VectorObject[Double] = RealVector(this.values.mapValues(_.middlePoint))

  final def split(ratios: Seq[Double], key: Option[String] = None): Seq[IntervalVector] = {
    val splitKey =
      if (key.isDefined) key.get
      else values.minBy { case (k, v) => -v.width }._1

    val splitComponent = this(splitKey).split(ratios)
    splitComponent.map { i => IntervalVector(this.values + (splitKey -> i))}
  }

  final def bisect(key: Option[String] = None): (IntervalVector, IntervalVector) = {
    val Seq(left, right) = this.split(Seq(1.0, 1.0), key)
    (left, right)
  }

}

/** Companion object for [[OSOL.Extremum.Core.Scala.Vectors.IntervalVector IntervalVector]] class */
object IntervalVector {

  /** Implicit converters for [[OSOL.Extremum.Core.Scala.Vectors.IntervalVector IntervalVector]] */
  object Converters {

    /** Converts `Iterable[(String, Interval)]` to [[OSOL.Extremum.Core.Scala.Vectors.IntervalVector IntervalVector]]
      *
      * @param v `(key, value)` pairs
      * @return [[OSOL.Extremum.Core.Scala.Vectors.IntervalVector IntervalVector]]
      */
    implicit def Iterable_to_IntervalVector(v: Iterable[(String, Interval)]): IntervalVector = IntervalVector(v)

    /** Converts `VectorObject[Double]` to [[OSOL.Extremum.Core.Scala.Vectors.IntervalVector IntervalVector]]
      *
      * @param v `VectorObject[Double]`
      * @return [[OSOL.Extremum.Core.Scala.Vectors.IntervalVector IntervalVector]]
      */
    implicit def VectorObject_to_IntervalVector(v: VectorObject[Interval]): IntervalVector = IntervalVector(v.values)

  }

  /** Create object from `(key, value)` pairs
    *
    * @param keyValuePairs target `(key, value)` pairs
    * @return vector composed of `keyValuePairs`
    */
  final def apply(keyValuePairs: (String, Interval)*): IntervalVector = new IntervalVector(keyValuePairs.toMap)
  /** Create object from `(key, value)` pairs
    *
    * @param keyValuePairs target `(key, value)` pairs
    * @return vector composed of `keyValuePairs`
    */
  final def apply(keyValuePairs: Iterable[(String, Interval)]): IntervalVector = IntervalVector(keyValuePairs.toSeq:_*)


}

File: Cores/DotNet/Vectors/IntervalVector.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Newtonsoft.Json.Linq;
using OSOL.Extremum.Cores.DotNet.Arithmetics;
using OSOL.Extremum.Cores.DotNet.Optimization;
using OSOL.Extremum.Cores.DotNet.Random;

namespace OSOL.Extremum.Cores.DotNet.Vectors
{
    public class IntervalVector : VectorObject<Interval>, IOptimizable<IntervalVector, Interval>
    {
        public IntervalVector(Dictionary<string, Interval> elements)
        {
            this.Elements = elements;
        }

        public IntervalVector(JObject json)
        {
            Dictionary<string, Interval> elements = new Dictionary<string, Interval>();
            foreach (var j in (JArray) json["IntervalVector"]["elements"])
            {
                elements.Add(j["key"].Value<string>(), new Interval((JObject) j["value"]));
            }

            this.Elements = elements;
        }

        public static implicit operator IntervalVector(Dictionary<string, Interval> elements) =>
            new IntervalVector(elements);

        public sealed override bool EqualsTo(VectorObject<Interval> that)
        {
            string[] keys_1 = this.Keys.ToArray();
            string[] keys_2 = that.Keys.ToArray();
            if (!(keys_1.All(k => keys_2.Contains(k)) && keys_2.All(k => keys_1.Contains(k))))
            {
                throw new VectorExceptions.DifferentKeysException(keys_1, keys_2);
            }
            else
            {
                return keys_1.All(k => this[k] == that[k]);
            }
        }

        public sealed override VectorObject<Interval> Add(VectorObject<Interval> that) =>
            new IntervalVector(this.ElementWiseOp(that, (x, y) => x + y));

        public sealed override VectorObject<Interval> AddImputeMissingKeys(VectorObject<Interval> that) =>
            new IntervalVector(this.ElementWiseOpImputeMissingKeys(that, (x, y) => x + y, defaultValue: 0.0));

        public sealed override VectorObject<Interval> Multiply(double coefficient) =>
            new IntervalVector(this.Elements.ToDictionary(x => x.Key, x => coefficient * x.Value));

        public sealed override VectorObject<Interval> Multiply(VectorObject<Interval> that) =>
            new IntervalVector(this.ElementWiseOp(that, (x, y) => x * y));

        public sealed override VectorObject<Interval> MultiplyImputeMissingKeys(VectorObject<Interval> that) =>
            new IntervalVector(this.ElementWiseOpImputeMissingKeys(that, (x, y) => x * y, defaultValue: 1.0));

        public sealed override VectorObject<Interval> Subtract(VectorObject<Interval> that) =>
            new IntervalVector(this.ElementWiseOp(that, (x, y) => x - y));

        public sealed override VectorObject<Interval> SubtractImputeMissingKeys(VectorObject<Interval> that) =>
            new IntervalVector(this.ElementWiseOpImputeMissingKeys(that, (x, y) => x - y, defaultValue: 0.0));

        public IntervalVector[] Split(double[] ratios, string key = null)
        {
            string splitKey = key;
            if (splitKey == null)
            {
                GoRN gorn = new GoRN();
                double minWidth = this.Elements.OrderBy(_ => -_.Value.Width).First().Value.Width;
                var smallestComponents = this.Elements
                    .Where(kvp => Math.Abs(kvp.Value.Width - minWidth) < Interval.MinWidth)
                    .Select(kvp => kvp.Key)
                    .ToList();
                splitKey = gorn.GetFromSeries(smallestComponents, 1, false).First();
            }
            var splitComponents = this[splitKey].Split(ratios);
            return splitComponents
                .Select(c =>
                {
                    var remaining = this.Elements
                        .Where(_ => !_.Key.Equals(splitKey));
                    remaining = remaining.Append(KeyValuePair.Create(splitKey, c));
                    return remaining;
                }).Select(v => new IntervalVector(v.ToDictionary(kvp => kvp.Key, kvp => kvp.Value)))
                .ToArray();
        }

        public Tuple<IntervalVector, IntervalVector> Bisect(string key = null)
        {
            var divided = this.Split(ratios: new [] {1.0, 1.0}, key: key);
            return Tuple.Create(divided[0], divided[1]);
        }

        public IntervalVector MoveBy(Dictionary<string, double> delta) =>
            this.AddImputeMissingKeys(
                new IntervalVector(delta.ToDictionary(_ => _.Key, _ => new Interval(_.Value)))).Elements;

        public IntervalVector Constrain(Dictionary<string, Tuple<double, double>> area)
        {
            var result = this.Elements
                .ToDictionary(
                    kvp => kvp.Key,
                    kvp =>
                    {
                        double min = double.NegativeInfinity, max = double.PositiveInfinity;
                        if (area.ContainsKey(kvp.Key))
                        {
                            min = area[kvp.Key].Item1;
                            max = area[kvp.Key].Item2;
                        }

                        return kvp.Value.Constrain(min, max);
                    });
            return new IntervalVector(result);
        }

        public double GetPerformance(Func<Dictionary<string, Interval>, Interval> f) => f(this.Elements).LowerBound;

        public VectorObject<double> ToBasicForm() =>
            new RealVector(this.Elements.ToDictionary(_ => _.Key, _ => _.Value.MiddlePoint));

        public override VectorObject<Interval> Union(params Tuple<string, Interval>[] vectors)
        {
            var allKeys = new List<string>();
            allKeys.AddRange(this.Keys);
            allKeys.AddRange(vectors.Select(_ => _.Item1));
            allKeys = allKeys.Distinct().ToList();
            return new IntervalVector(allKeys.ToDictionary(k => k, k =>
            {
                var targetKey = vectors.Where(_ => _.Item1.Equals(k)).ToList();
                return targetKey.Count == 0 ? this[k] : targetKey.First().Item2;
            }));
        }
        
        public override Dictionary<string, double> DistanceFromArea(Dictionary<string, Tuple<double, double>> area)
        {
            return area.ToDictionary(kvp => kvp.Key, kvp =>
            {
                var min = kvp.Value.Item1;
                var max = kvp.Value.Item2;

                var i = this[kvp.Key];
                Func<double, double> dist = v =>
                {
                    if (v < min) return min - v;
                    else
                    {
                        if (v > max) return v - max;
                        else return 0.0;
                    }
                };
                return Math.Max(dist(i.LowerBound), dist(i.UpperBound));
            });
        }

        public JObject ConvertToJson()
        {
            JObject json = new JObject();
            json["IntervalVector"] = new JObject();
            JArray elements = new JArray();
            foreach (var kvp in this.Elements)
            {
                JObject temp = new JObject();
                temp["key"] = kvp.Key;
                temp["value"] = kvp.Value.ConvertToJson();
                elements.Add(temp);
            }

            json["IntervalVector"]["elements"] = elements;
            return json;
        }

    }
}